keps/sig-api-machinery/2887-openapi-enum-types/README.md

```diff
@@ -0,0 +1,830 @@
+type ExampleResource struct {
+    FooField FooType
+}
+
+```
+
+### Enum Tag
+The Enum Tag is the annotation `// +enum` that must be in the closest comment blocks
+of an enum type following Enum Pattern.
+
+```go
+package internal
+
+// FooType define sa type alias, underlying type must be string
+// +enum
+type FooType string
+```
+
+An actual example looks like
+```go
+package core
+
+// Protocol defines network protocols supported for things like container ports.
+// +enum
+type Protocol string
+```
+
+### Detect Possible Enum Types
+
+A new API rule will be added to `kube-openapi` that warns about detected Enum Pattern
+without the enum tag. If, in a rare situation, the detected type is not an enum, the user can
+add an exception in `/api/api-rules/`.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Not all types following the Enum Pattern are actually enum types. Here is a counter-example.
+```go
+package core
+
+// StorageMedium defines ways that storage can be allocated to a volume.
+type StorageMedium string
+
+// These are the valid value for StorageMedium
+const (
+  StorageMediumDefault         StorageMedium = ""           // use whatever the default is for the node
+  StorageMediumMemory          StorageMedium = "Memory"     // use memory (tmpfs)
+  StorageMediumHugePages       StorageMedium = "HugePages"  // use hugepages
+  StorageMediumHugePagesPrefix StorageMedium = "HugePages-" // prefix for full medium notation HugePages-<size>
+)
+```
+
+Here, `StorageMedium` can have infinite number of possible values, which disqualify it as an enum type.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The syntax proposed in this KEP is different from existing implementation of kube-builder.
+Namely, kube-builder requires possible values in the annotation itself, and is treated
+as a validation.
+```go
+package example
+
+// Alias
+// +kubebuilder:validation:Enum=Lion;Wolf;Dragon
+type Alias string
+```
+The two approaches are compatible, but eventually they will have to be unified.
```

Comment:
@jiahuif you linked me to a location where you were concerned about the impact of this change on generated code.  If someone is using the kubebuilder mechanism and switches to the enum you're providing here, does that trigger the same condition you were concerned about?

Reply:
No. For kube-builder, the existing enum annotation adds the check to a validation webhook. The generated code does not yet have the enum tag. In this KEP, we are to add enum support to the core types, changes to the ecosystem can be done afterwards.

Reply:
Correction: I've updated the section about kube-builder. I will add support of the new syntax to kube-builder too.